Source/Details/ASPINRemoteImageDownloader.m

```diff
@@ -116,56 +115,67 @@ + (ASPINRemoteImageDownloader *)sharedDownloader NS_RETURNS_RETAINED
 {
-  NSAssert(sharedDownloader == nil, @"Singleton has been created and session can no longer be configured.");
-  NSAssert1(sharedPINRemoteImageManager == nil, @"An instance of %@ has been set. Either configuration or preconfigured image manager can be set at a time and only once.", [[sharedPINRemoteImageManager class] description]);
-  __unused PINRemoteImageManager *sharedManager = [self sharedPINRemoteImageManagerWithConfiguration:configuration preconfiguredPINRemoteImageManager:nil];
+  PINRemoteImageManager *sharedManager = [self PINRemoteImageManagerWithConfiguration:configuration imageCache:nil];
+  [self setSharedPreconfiguredRemoteImageManager:sharedManager];
 }
 
-+ (void)setSharedPreconfiguredRemoteImageManager:(nullable PINRemoteImageManager *)preconfiguredPINRemoteImageManager
++ (void)setSharedImageManagerWithConfiguration:(nullable NSURLSessionConfiguration *)configuration
+                                    imageCache:(nullable id<PINRemoteImageCaching>)imageCache
+{
+  PINRemoteImageManager *sharedManager = [self PINRemoteImageManagerWithConfiguration:configuration imageCache:imageCache];
+  [self setSharedPreconfiguredRemoteImageManager:sharedManager];
+}
+
+static dispatch_once_t shared_init_predicate;
+
++ (void)setSharedPreconfiguredRemoteImageManager:(PINRemoteImageManager *)preconfiguredPINRemoteImageManager
 {
   NSAssert(sharedDownloader == nil, @"Singleton has been created and session can no longer be configured.");
   NSAssert1(sharedPINRemoteImageManager == nil, @"An instance of %@ has been set. Either configuration or preconfigured image manager can be set at a time and only once.", [[sharedPINRemoteImageManager class] description]);
-  __unused PINRemoteImageManager *sharedManager = [self sharedPINRemoteImageManagerWithConfiguration:nil preconfiguredPINRemoteImageManager:preconfiguredPINRemoteImageManager];
+
+  dispatch_once(&shared_init_predicate, ^{
+    sharedPINRemoteImageManager = preconfiguredPINRemoteImageManager;
+  });
 }
 
-+ (PINRemoteImageManager *)sharedPINRemoteImageManagerWithConfiguration:(NSURLSessionConfiguration *)configuration preconfiguredPINRemoteImageManager:(PINRemoteImageManager *)preconfiguredPINRemoteImageManager
++ (PINRemoteImageManager *)PINRemoteImageManagerWithConfiguration:(nullable NSURLSessionConfiguration *)configuration imageCache:(nullable id<PINRemoteImageCaching>)imageCache
 {
-  NSAssert(!(configuration != nil && preconfiguredPINRemoteImageManager != nil), @"Either configuration or preconfigured image manager can be set at a time.");
-  static dispatch_once_t onceToken;
-  dispatch_once(&onceToken, ^{
-
-    if (preconfiguredPINRemoteImageManager) {
-      sharedPINRemoteImageManager = preconfiguredPINRemoteImageManager;
-    } else {
+  PINRemoteImageManager *manager = nil;
+  // Check that Carthage users have linked both PINRemoteImage & PINCache by testing for one file each
+  if (!(NSClassFromString(@"PINRemoteImageManager"))) {
```

Comment:
We should put that behind a debug flag.

Reply:
OK

Reply:
I wondered as I reviewed before why this is the case? cc @garrettmoon

Reply:
I think I hit the error condition once, and didn't get the exception, and saw that, and thought it would be helpful, and it didn't have anything really to do with `PIN_ANIMATED_AVAILABLE` specifically, so I moved it out. *takes a breath*

I can see putting it behind `#if DEBUG`, although it's just going to crash either way and for a check that is guaranteed to only ever run once it doesn't seem like two `NSClassFromString`s is that expensive. But it's of limited use without a debugger attached (you would get a friendlier crash report from Apple if your released app hit here, but then presumably your app is crashing every time so ...... you hopefully have it figured out)